### PR TITLE
TextAdventureGameTeam1-Assignment3

### PR DESCRIPTION
**Pull request by a-wilkes**

### CODE REVIEW

File: docs/assignment2.md [hunk 1]
```diff
@@ -228,7 +228,7 @@ Attributes
 Author(s): `Anthony Wilkes, Ajay Hitendra Mota`
 
 The figure representing the UML object diagram is as follows :-
-![Object Diagram](./assets/object_diagram.PNG "State machine diagram for the Engine class")
+![Object Diagram](assets/A02/object_diagram.PNG "State machine diagram for the Engine class")
 
 In the above diagram, the system is in a state wherein the game has been loaded successfully and the gamer just inputted the command `eat apple` into the system. The description of the various objects and the states of their variables are as follows :-
```
Comment (ghost) [inline docs/assignment2.md:233]: Looks good

File: docs/assignment2.md [hunk 2]
```diff
@@ -253,7 +253,7 @@ Author(s): `Anthony Wilkes, Ajay Hitendra Mota`
 In this section, we will describe two components of the system that hold different states. There are the **Engine** class and the **UIHandler** class.
 
 ### Engine
-![Engine State Machine](./assets/engine_state.png "State machine diagram for the Engine class")
+![Engine State Machine](assets/A02/engine_state.png "State machine diagram for the Engine class")
```
Comment (ghost) [inline docs/assignment2.md:256]: Looks good

File: docs/assignment2.md [hunk 3]
```diff
@@ -253,7 +253,7 @@ Author(s): `Anthony Wilkes, Ajay Hitendra Mota`
...
 
 The **Engine** class will be the main, top-level control point for the system as a whole. On system start the **Engine** will load the list of available games that the user can play to present them in a menu. This list will be determined by the **Initializer** as it is the **Initializer** that should be responsible for interactions with game files. If the **Initializer** fails at any point during this process, it should return a reason why and this error response should be printed to the console before the system exits.
```
Comment (ghost) [inline docs/assignment2.md:258]: Nice.

File: docs/assignment2.md [hunk 4]
```diff
@@ -299,7 +299,7 @@ In this section, we will describe two crucial parts of the system during its fun
 ### Game Initialization Sequence
 
 The UML diagram for the interaction is as follows :-
-![State machine diagrams](./assets/init_sequence.PNG "State machine diagram 1")
+![State machine diagrams](assets/A02/init_sequence.PNG "State machine diagram 1")
 
 The game initialization sequence is divided into two main parts. The upper half of the diagram shows the initialization of the game with the correct objects, their lifelines, their types and the right functions/values used as messages. The lower half shows the two alternate scenarios in prescriptive detail till the point in time that the game just starts its execution. The description of the two main phases is as follows :-
```
Comment (ghost) [inline docs/assignment2.md:304]: Replace till with untill

File: docs/assignment2.md [hunk 5]
```diff
@@ -299,7 +299,7 @@ In this section, we will describe two crucial parts of the system during its fun
...
 - Pre-Initialization: The engine is the main controller of the game invokes the initializer object bypassing the game path of the selected user's choice from the UIHandler. This is abstracted away in this sequence diagram. The main initialization occurs from this step as follows :-
```
Comment (ghost) [inline docs/assignment2.md:305]: bypassing > by passing

File: src/main/java/command/ChangeLocation.java [hunk 6]
```diff
@@ -3,46 +3,92 @@
...
 
-    public String apply(final Npc object, final GameEntities ge) {
-        return "";
     }
 
-    public String apply(final Obstacle object, final GameEntities ge) {
-        return "";
-    }
+    private void
+    removeEntityFromArea(final Locatable targetEntity, final GameEntities ge)
```
Comment (ghost) [inline src/main/java/command/ChangeLocation.java:78]: Looks good

File: src/main/java/command/Describe.java [hunk 7]
```diff
@@ -3,45 +3,23 @@
...
 
-        Entity e = ge.findEntityOrElse(object);
+        final Entity e = ge.getEntityOrDefault(object);
+        final Area localArea = ge.getAreaOrDefault(ge.getPlayer().getCurrentLocation());
 
-        if (!e.isActive()) { return String.format("You cannot see '%s'.", object); }
+        if (!e.isActive()
+                || !(ge.getPlayer().hasInInventory(object)
+                    || (object.equals(ge.getPlayer().getName()))
+                    || localArea.hasInInventory(object)
```
Comment (ghost) [inline src/main/java/command/Describe.java:17]: ...

File: src/main/java/command/TakeItem.java [hunk 8]
```diff
@@ -3,61 +3,40 @@
...
-        Area area = ge.getAreaOrElse(parent.getCurrentLocation());
-
-        if (!area.hasInInventory(object)) { return String.format("You cannot pick up '%s'", object); }
+        if (!area.hasInInventory(object)) { return String.format("You cannot see '%s' here.", object); }
 
         area.removeFromInventory(object);
 
-        StringBuilder result = new StringBuilder(object + " -> " + parent.getName());
+        StringBuilder result = new StringBuilder(parent.getName() + " picked up " + object);
+
```
Comment (ghost) [inline src/main/java/command/TakeItem.java:24]: Nice

File: src/main/java/cork/engine_state/HomeScreen.java [hunk 9]
```diff
@@ -0,0 +1,13 @@
+package cork.engine_state;
```
Comment (ghost) [inline src/main/java/cork/engine_state/HomeScreen.java:1]: clean++

File: src/main/java/command/Fight.java [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+        else if (parentStrength.compareTo(opponentStrength) > 0)
+        {
+            targetEntity.setStat(REDUCTION_STAT, 0);
+            targetEntity.setActive(false);
+
+            StringBuilder result = new StringBuilder(String.format("%s is destroyed.", object));
+
+            final Area location = ge.getAreaOrDefault(ge.getPlayer().getCurrentLocation());
+            for (final String item : targetEntity.getInventory())
+            {
```
Comment (ghost) [inline src/main/java/command/Fight.java:67]: less \n pls

**Review by ghost — COMMENTED** [review]
Overal good coding, not much to remark on! good job.

File: out/artifacts/cork_jar/game_template/areas.json [hunk 11]
```diff
@@ -0,0 +1,28 @@
+[
+    {
+        "active": true,     // Current status - this should always be true
```
Comment (TorsteinTh) [inline out/artifacts/cork_jar/game_template/areas.json:3]: Why does some attributes in .json files have comments and not all?

File: src/main/java/command/Describe.java [hunk 12]
```diff
@@ -3,45 +3,23 @@
...
     @Override
     public String apply(String object, final GameEntities ge) {
         if (object.equals("")) { object = ge.getPlayer().getCurrentLocation(); }
 
-        Entity e = ge.findEntityOrElse(object);
+        final Entity e = ge.getEntityOrDefault(object);
+        final Area localArea = ge.getAreaOrDefault(ge.getPlayer().getCurrentLocation());
 
-        if (!e.isActive()) { return String.format("You cannot see '%s'.", object); }
+        if (!e.isActive()
```
Comment (TorsteinTh) [inline src/main/java/command/Describe.java:14]: longest if  test i've ever seen

File: src/main/java/command/CommandFactory.java [hunk 13]
```diff
@@ -1,22 +1,25 @@
...
+import entity.Entity;
 import deserialiser.CommandBlueprint;
+import entity.Locatable;
+import entity.Obstacle;
 
 public final class CommandFactory {
     private CommandFactory() {}
 
-    public static Command createCommand(final CommandBlueprint cmd, final IInteractable parent) {
+    public static ICommand createCommand(final CommandBlueprint cmd, final Entity parent) {
```
Comment (TorsteinTh) [inline src/main/java/command/CommandFactory.java:11]: great :)

File: src/main/java/command/ChangeLocation.java [hunk 14]
```diff
@@ -3,46 +3,92 @@
 import dictionary.GameEntities;
 import entity.*;
 
-import java.util.Set;
+import java.util.Map;
 
-public class ChangeLocation extends Command {
-    private IInteractable parent;
-    private Set<String> args;
+public class ChangeLocation implements ICommand {
```
Comment (TorsteinTh) [inline src/main/java/command/ChangeLocation.java:8]: looks great

**Review by TorsteinTh — COMMENTED** [review]
Overall looks great! Mainly json files that's added, Nothing much to say about that other than some have comments, and some don't. Would like to see some comments on some of the functions, to discribe their functionality. Great work :)

**Review by Ece-Doganer — COMMENTED** [review]
Dear team,

Great job on your project. The code is beautiful, its design is impressive. The updates from A2 to A3 make sense. Nice usage of design patterns.
My message from A2 stands: very good structured code, clean. Great solutions.
Now on to the nitpicking.

**Gameplay suggestions:**
- player doesn't know which commands are available.
- Player doesn’t know which directions are available.
- Most of the input of the player will result in an error message, as this is how text-adventure games are played. Your default messages in this perspective are unfriendly and short. Since they are hardcoded, it would be nice if they at least are helpful. Additionally, it might have been an idea back at the design phase, to incorporate storytelling through "error" messages. Per player action you would have text for the success or failure of that method. For example: "You take a big bite out of the apple and savour its sweetness" or, "The apple is made out of wax, you consider it inedible.". Additionally: "You cannot do that"-message is repeated for different actions. Could have given more logical clues here, breaking the repetitiveness of this message for betteruser-experience.
- It could have been an idea to make different words available for the same method. For example: take, pick-up, pick, grab, get. But knowing the command for the manual at the start of a game, would already help.
- Home screen's continue option, for a player, builds anticipation. Could have immediately had game menu options on there. I appreciate you didn't.

**General suggestions:**
Some lines in your code are still longer than 180 chars.

**Specific suggestions:**
cork.engine_state.Engine:
16: + Nice use of naming conventions for static final fields.
75: + I appreciate the multiple try and catch blocks in different methods, as it is now more clear per method call what error cases have been taken into account.
111, 129: - This variable reads a bit messy, and clarity is king. I would suggest multiple variables, to make reading the steps more straightforward. Or less newlines.
39, 143: - engine.run and engine.runGame do not make clear in their naming what their difference is. Ward Cunningham: “You know you are working on clean code when each routine you read turns out to be pretty much what you expected. You can call it beautiful code when the code also makes it look like the language was made for the problem.”

cork.engine_state:
Great job on the design following states. Very clean.

cork.Initializer:
144: - Closing bracket could follow to return.

command.Fight:
43: - Too much happening in one line.

command.Describe:
14: - Very messy to make out what is happening in terms of parentheses.

command.ChangeState:
15: - These types of long if-statements cognitively overload the reader. It would be more welcoming to think along for fellow (future) team-members, to make smaller steps.
It might look clean in one line, but code is communication with team members first, and communication with machine second. Aim to make it easy for other people to engage and enhance.

Finally: Great job guys. Really well done.

**Review by malekkn — COMMENTED** [review]
The application turned out as good as it was expected. Good use of templates and impressive collection of games. The design patterns used in an appropriate areas. Feedback from the previous release was nicely processed. The API exposed between packages is nice and simple enough from the client code point of view. In general, The source code is easy to understand and hard to criticize. Here is some Feedback:

- Some lines are really long and can be broken down as mentioned by other reviewers.
- There is some duplicate code between three classes: ChangeLocation: 43-50, ChangeStat: 26-33 and TakeItem: 31-38.
- Similar repetition can be found within one class such as lines 45-50 and 65-70 in Fight. 

Congratulations on this work!

File: docs/assignment3.md [hunk 15]
```diff
@@ -1 +1,709 @@
 # Assignment 3
+
```
Comment (Ferch42) [inline docs/assignment3.md:2]: Very well structured, well thought out and designed project. The code is very clean and the diagrams are clear and compact. The explanations are also very clear and understandable. Congratulations! :)

**Comment by Ferch42** [comment]
Very good job, congratulations for your effort and dedication

**Pull request merged**